### DISCUSSION TITLE
How to add postgres library before compilation

### DISCUSSION THREAD

**Discussion by corvusrabus** [cross-rs/cross#1163] [Q&A]
Hi,

I am trying to cross-compile from my x86 CPU to the target aarch64-unknown-linux-gnu and my code needs to be linked with libpq (the Postgres SQL library).
The error I get is: /usr/bin/ld: cannot find -lpq

I have tried multiple things already and none of them have worked. I think the biggest confusion for me is that the docker image that cross runs is an x86 image but the binary needs to be linked with an aarch64 version of libpq.

Any help would be greatly appreciated.

**Comment by Alexhuszagh** [comment]
See the wiki on [working with external dependencies](https://github.com/cross-rs/cross/wiki/FAQ#external-dependencies).

In this case, you'd likely want in `Cross.toml`:

```toml
[target.aarch64-unknown-linux-gnu]
pre-build = [
  "dpkg --add-architecture $CROSS_DEB_ARCH", 
  "apt-get update && apt-get --assume-yes install libpq-dev:$CROSS_DEB_ARCH"
]
```

And then it should just work.

**Reply by corvusrabus** [reply]
Hi Alex, thanks for the answer.
 I am sorry that I wasn't more specific about what I have tried before. I got stuck on this some months ago and then decided it was not worth the effort and started using an x86 box instead.

I have tried the suggestion before and it does not work for me and I get the exact same error( I do see though that cross instalsl libpq-dev during the build process). I think that the problem is that the linker does not automatically find libpq because it only looks for x86 libraries since the docker image is an x86 image. 

I am sorry if I am expressing myself badly. I have very little experience with cross-compilation and was super happy to find cross which worked straight out of the box for everything until I needed to bind the Postgres library.

**Reply by Alexhuszagh** [reply]
That's odd, it should be using the correct library. Is it doing this in a proc macro? Can you give me a more concrete example, and I'd be happy to help more. It doesn't have to be any proprietary code, just a minimal example of what's failing and I'd be happy to help in whatever way I can.

There's a few things we do that should help it find the library normally:
1. We use the default install, so it should be installed by default in the compiler search path.
2. We export `PKG_CONFIG_PATH="/usr/lib/aarch64-linux-gnu/pkgconfig/:${PKG_CONFIG_PATH}"`, which means that any libraries using `pkg-config` should be able to find the dependencies.

A more likely cause is that you are using a proc macro somewhere in the dependency list, which might cause it to be attempting to use the host compiler. I'll be able to debug better if I can get a more concrete example.

**Reply by corvusrabus** [reply]
Thanks for offering help Alex.

You're probably right about the proc macro problem. I think the absolute minimal example would be to just include the dependencies I need into an otherwise empty crate. Unfortunately, I don't get that to compile right now and the error message is different. I need the following crates:
`rust_decimal = { version =  "1.26.1", features = ["db-diesel2-postgres"] }
diesel = { version = "2.0.2",  features = ["chrono","postgres_backend"] }
`
This now gives me the following error:

> error: failed to run custom build command for `serde v1.0.148`

Caused by:
  process didn't exit successfully: `/target/debug/build/serde-f67f74b60d097ec1/build-script-build` (exit status: 1)
  --- stderr
  /target/debug/build/serde-f67f74b60d097ec1/build-script-build: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /target/debug/build/serde-f67f74b60d097ec1/build-script-build)
  /target/debug/build/serde-f67f74b60d097ec1/build-script-build: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by /target/debug/build/serde-f67f74b60d097ec1/build-script-build)
  /target/debug/build/serde-f67f74b60d097ec1/build-script-build: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by /target/debug/build/serde-f67f74b60d097ec1/build-script-build)
> 

Both crates I need depend on the crate [pq-sys](https://github.com/sgrif/pq-sys) which is a Rust binding to libpq. I read a bit more about that crate and it depends on [pkg_config](https://docs.rs/pkg-config/latest/pkg_config/) that explicitly says that it does not usually support cross-compilation

**Reply by Emilgardis** [reply]
the glibc error is temporary, see https://github.com/cross-rs/cross/issues/724

for the postgres dep, try adding 

```toml
[dependencies]
pq-sys = {version = "0.4", features = ["pkg-config"]}
```

to enable `pkg-config`

**Reply by corvusrabus** [reply]
Hi Emil, thanks for helping as well. Your suggestion makes the bare-bones example I described above work!
In my actual project I get this error now:

>  /usr/bin/ld: skipping incompatible /usr/lib/aarch64-linux-gnu/libpq.so when searching for -lpq
          /usr/bin/ld: skipping incompatible /usr/lib/aarch64-linux-gnu/libpq.a when searching for -lpq
          /usr/bin/ld: cannot find -lpq
          /usr/bin/ld: skipping incompatible /usr/lib/aarch64-linux-gnu/libgcc_s.so.1 when searching for libgcc_s.so.1
          collect2: error: ld returned 1 exit status

I tried several times (with cargo cleans in between) with PKG_CONFIG_ALLOW_CROSS=1 and without PKG_CONFIG_ALLOW_CROSS

**Reply by Emilgardis** [reply]
Do you get that with `Cross.toml` containing

```toml
[target.aarch64-unknown-linux-gnu]
pre-build = [
  "dpkg --add-architecture $CROSS_DEB_ARCH", 
  "apt-get update && apt-get --assume-yes install libpq-dev:$CROSS_DEB_ARCH"
]
```

**Reply by corvusrabus** [reply]
Sorry that I didn't specify that. Yes, my Cross.toml contains that.

**Reply by Alexhuszagh** [reply]
@Emilgardis I'm pretty sure it's the same thing as discussed here, since it's using a proc macro and then running it on the x86_64 host. I'm trying to find a solution for proc macros just to confirm everything is done properly. The issue what I'm running into is proc macros are run on the host, so in this case, `x86_64`, but then the target they are compiled for is `aarch64`, and I'm not sure even if the code compilers if it is correct. **I believe it is** as long as it uses a shared library and then it requires the `libql-dev:arm64` to run but not to compile, but I believe this will fail if it's a static library. I'm looking into this currently.

https://github.com/cross-rs/cross/issues/605#issuecomment-1328316874

**Reply by Alexhuszagh** [reply]
> Sorry that I didn't specify that. Yes, my Cross.toml contains that.

Ok we just discussed this in the matrix room: you will need both `libpq-dev` and `libpq-dev:arm64`, and we'll document this on the wiki shortly. The reason is the proc macro compiles for the host, but has no access to the final binary (only the `TokenStream`), so you will need it both for the host and the target.

**Comment by Alexhuszagh** [comment]
If you are using a procedural macro, you will need the dependencies for the proc macro on the host, and the dependencies for the post-expansion crate on the target. In this case, both are `libpq-dev`, so you will need `libpq-dev` and `libpq-dev:arm64`. We've now documented this on the [wiki](https://github.com/cross-rs/cross/wiki/FAQ#procedural-macros).

Let me know if this fixes the issue.

**Reply by corvusrabus** [reply]
Hi Alex. Unfortunately, the docker image won't build if I do what's written in the wiki. Apt complains that libpq-dev and libpq-dev:arm64 are conflicting packages. I have tried 
`[target.aarch64-unknown-linux-gnu]
pre-build = [
    "dpkg --add-architecture $CROSS_DEB_ARCH",
    "apt-get update && apt-get --assume-yes install libpq-dev:amd64 libpq-dev:$CROSS_DEB_ARCH"
]`
and I get the following error from apt

> #0 14.75 Some packages could not be installed. This may mean that you have
#0 14.75 requested an impossible situation or if you are using the unstable
#0 14.75 distribution that some required packages have not yet been created
#0 14.75 or been moved out of Incoming.
#0 14.75 The following information may help to resolve the situation:
#0 14.75 
#0 14.75 The following packages have unmet dependencies:
#0 14.82  libpq-dev : Conflicts: libpq-dev:arm64 but 12.12-0ubuntu0.20.04.1 is to be installed
#0 14.82  libpq-dev:arm64 : Conflicts: libpq-dev but 12.12-0ubuntu0.20.04.1 is to be installed
#0 14.84 E: Unable to correct problems, you have held broken packages.
------
ERROR: failed to solve: executor failed running [/bin/sh -c eval "${CROSS_CMD}"]: exit code: 100

**Reply by Alexhuszagh** [reply]
Ah this is why:

- [liqpq-dev:arm64](https://packages.ubuntu.com/bionic/arm64/libpq-dev/filelist)
- [liqpd-dev:amd64](https://packages.ubuntu.com/bionic/amd64/libpq-dev/filelist)

There's a solution using `apt download` and `dpkg -x`, and is probably worth mentioning on the wiki. I'll try to get a working example. You will need to modify the include paths if you do. Specifically, all these files overlap:

```
/usr/bin/pg_config
/usr/include/postgresql/internal/c.h
/usr/include/postgresql/internal/libpq-int.h
/usr/include/postgresql/internal/libpq/pqcomm.h
/usr/include/postgresql/internal/port.h
/usr/include/postgresql/internal/postgres_fe.h
/usr/include/postgresql/internal/pqexpbuffer.h
/usr/include/postgresql/libpq-events.h
/usr/include/postgresql/libpq-fe.h
/usr/include/postgresql/libpq/libpq-fs.h
/usr/include/postgresql/pg_config.h
/usr/include/postgresql/pg_config_ext.h
/usr/include/postgresql/pg_config_manual.h
/usr/include/postgresql/pg_config_os.h
/usr/include/postgresql/postgres_ext.h
/usr/share/doc/libpq-dev/changelog.Debian.gz
/usr/share/doc/libpq-dev/copyright
/usr/share/man/man1/pg_config.1.gz
```

Installing them local to `/usr/aarch64-linux-gnu` should suffice, which would probably mean `dpkg -x` to some temp directory, then doing something like:

```bash
dirs=(bin lib include)
for dir in "${dirs[@]}"; do
    for f in "${td}/usr/${dir}"/*; do
        base=$(basename "${f}")
        cp -r "${f}" "${CROSS_SYSROOT}/${dir}/${base}"
    done
done
```

Note that this isn't tested code and might not work, but the general idea is there: install every `aarch64` library to the cross sysroot, and every host library to the normal sysroot.

**Comment by Barugon** [comment]
> See the wiki on [working with external dependencies](https://github.com/cross-rs/cross/wiki/FAQ#external-dependencies).
> 
> In this case, you'd likely want in `Cross.toml`:
> 
> ```toml
> [target.aarch64-unknown-linux-gnu]
> pre-build = [
>   "dpkg --add-architecture $CROSS_DEB_ARCH", 
>   "apt-get update && apt-get --assume-yes install libpq-dev:$CROSS_DEB_ARCH"
> ]
> ```
> 
> And then it should just work.

Sweet! this totally solved my issue [here](https://github.com/cross-rs/cross/discussions/1318), although I had to add some timezone stuff.

```toml
[target.aarch64-unknown-linux-gnu]
pre-build = [
  "dpkg --add-architecture $CROSS_DEB_ARCH",
  "ln -snf /usr/share/zoneinfo/America/Los_Angeles /etc/localtime && echo America/Los_Angeles > /etc/timezone",
  "apt-get update && apt-get --assume-yes install libgdal-dev:$CROSS_DEB_ARCH",
]
```

**Reply by Barugon** [reply]
Ah, I spoke too soon.  The version of gdal I get is 2.6 but I need 3.6.  :disappointed:

**Reply by Emilgardis** [reply]
you'd have to install it yourself, ubuntu 18.04, 20.04 and 22.04 have <3.6

**Reply by Barugon** [reply]
Any suggestions for a base container?

**Reply by Emilgardis** [reply]
answered [here](https://github.com/cross-rs/cross/discussions/1318#discussioncomment-6806789)
